wgpu-core/src/adapter.rs

```diff
@@ -0,0 +1,246 @@
+                not(any(target_os = "ios", target_os = "macos")),
+                feature = "gfx-backend-vulkan"
+            ))]
+            Adapter::Vulkan(adapter) => {
+                log::info!("Adapter Vulkan {:?}", adapter.raw.info);
+                backend::Vulkan::hub(global).adapters.register_identity(
+                    backend_ids.vulkan.unwrap(),
+                    adapter,
+                    &mut token,
+                )
+            }
+            #[cfg(any(target_os = "ios", target_os = "macos"))]
+            Adapter::Metal(adapter) => {
+                log::info!("Adapter Metal {:?}", adapter.raw.info);
+                backend::Metal::hub(global).adapters.register_identity(
+                    backend_ids.metal.unwrap(),
+                    adapter,
+                    &mut token,
+                )
+            }
+            #[cfg(windows)]
+            Adapter::Dx12(adapter) => {
+                log::info!("Adapter Dx12 {:?}", adapter.raw.info);
+                backend::Dx12::hub(global).adapters.register_identity(
+                    backend_ids.dx12.unwrap(),
+                    adapter,
+                    &mut token,
+                )
+            }
+            #[cfg(windows)]
+            Adapter::Dx11(adapter) => {
+                log::info!("Adapter Dx11 {:?}", adapter.raw.info);
+                backend::Dx11::hub(global).adapters.register_identity(
+                    backend_ids.dx11.unwrap(),
+                    adapter,
+                    &mut token,
+                )
+            }
+        }
+    }
+}
+
+pub enum AdapterInputs<'a, I> {
+    IdSet(&'a [I], fn(&I) -> Backend),
+    Mask(BackendBit, fn() -> I),
+}
+
+impl<I: Clone> AdapterInputs<'_, I> {
+    fn find(&self, b: Backend) -> Option<I> {
+        match *self {
+            AdapterInputs::IdSet(ids, ref fun) => ids.iter().find(|id| fun(id) == b).cloned(),
+            AdapterInputs::Mask(bits, ref fun) => {
+                if bits.contains(b.into()) {
+                    Some(fun())
+                } else {
+                    None
+                }
+            }
+        }
+    }
+}
+
+bitflags::bitflags! {
+    #[repr(transparent)]
+    #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+    pub struct BackendBit: u32 {
+        const VULKAN = 1 << Backend::Vulkan as u32;
+        const GL = 1 << Backend::Gl as u32;
+        const METAL = 1 << Backend::Metal as u32;
+        const DX12 = 1 << Backend::Dx12 as u32;
+        const DX11 = 1 << Backend::Dx11 as u32;
+        /// Vulkan + METAL + DX12
+        const PRIMARY = Self::VULKAN.bits | Self::METAL.bits | Self::DX12.bits;
+        /// OpenGL + DX11
+        const SECONDARY = Self::GL.bits | Self::DX11.bits;
+    }
+}
+
+impl From<Backend> for BackendBit {
+    fn from(backend: Backend) -> Self {
+        BackendBit::from_bits(1 << backend as u32).unwrap()
+    }
+}
+
+pub(crate) struct BackendIds<F: IdentityFilter<AdapterId>> {
```

Comment:
I made this to move around `id_vulkan` and such around between functions.
Honestly I have a feeling that this is stupid, I hope there is an alternative to this.